src/display/annotation_layer.js

```diff
@@ -38,341 +38,515 @@ var CustomStyle = displayDOMUtils.CustomStyle;
+  function WidgetAnnotationElement(parameters) {
+    AnnotationElement.call(this, parameters);
   }
 
-  function update(viewport, div, annotations) {
-    for (var i = 0, ii = annotations.length; i < ii; i++) {
-      var data = annotations[i];
-      var element = div.querySelector(
-        '[data-annotation-id="' + data.id + '"]');
-      if (element) {
-        CustomStyle.setProp('transform', element,
-          'matrix(' + viewport.transform.join(',') + ')');
+  Util.inherit(WidgetAnnotationElement, AnnotationElement, {
+    /**
+     * Render the widget annotation's HTML element in the empty container.
+     *
+     * @public
+     * @memberof WidgetAnnotationElement
+     */
+    render: function WidgetAnnotationElement_render() {
+      var content = document.createElement('div');
+      content.textContent = this.data.fieldValue;
+      var textAlignment = this.data.textAlignment;
+      content.style.textAlign = ['left', 'center', 'right'][textAlignment];
+      content.style.verticalAlign = 'middle';
+      content.style.display = 'table-cell';
+
+      var font = (this.data.fontRefName ?
+        this.page.commonObjs.getData(this.data.fontRefName) : null);
+      this._setTextStyle(content, font);
+
+      this.container.appendChild(content);
+      return this.container;
+    },
+
+    /**
+     * Apply text styles to the text in the element.
+     *
+     * @private
+     * @param {HTMLDivElement} element
+     * @param {Object} font
+     * @memberof WidgetAnnotationElement
+     */
+    _setTextStyle:
+        function WidgetAnnotationElement_setTextStyle(element, font) {
+      // TODO: This duplicates some of the logic in CanvasGraphics.setFont().
+      var style = element.style;
+      style.fontSize = this.data.fontSize + 'px';
+      style.direction = (this.data.fontDirection < 0 ? 'rtl': 'ltr');
+
+      if (!font) {
+        return;
       }
+
+      style.fontWeight = (font.black ?
+        (font.bold ? 'bolder' : 'bold') :
+        (font.bold ? 'bold' : 'normal'));
+      style.fontStyle = (font.italic ? 'italic' : 'normal');
+
+      // Use a reasonable default font if the font doesn't specify a fallback.
+      var fontFamily = font.loadedName ? '"' + font.loadedName + '", ' : '';
+      var fallbackName = font.fallbackName || 'Helvetica, sans-serif';
+      style.fontFamily = fontFamily + fallbackName;
     }
-    div.removeAttribute('hidden');
-  }
+  });
+
+  return WidgetAnnotationElement;
+})();
 
+var AnnotationLayer = (function AnnotationLayerClosure() {
   return {
-    render: render,
-    update: update
+    /**
+     * Render a new annotation layer with all annotation elements.
+     *
+     * @public
+     * @param {PageViewport} viewport
+     * @param {PDFPage} page
+     * @param {HTMLDivElement} div
+     * @param {Array} annotations
+     * @param {IPDFLinkService} linkService
+     * @memberof AnnotationLayer
+     */
+    render: function AnnotationLayer_render(viewport, div, annotations, page,
```

Comment:
Would it perhaps be nicer, and more in line with `text_layer.js`, to instead replace these parameters with an object containing the necessary properties?